blocks/datasets/streams.py

```diff
@@ -314,6 +314,33 @@ def _cache(self):
             cache.extend(data)
 
 
+class SortMapping(object):
+    """Callable class for creating sorting mappings.
+
+    This class can be used to create a callable that can be used by the
+    :class:`DataStreamMapping` constructor.
+
+    Parameters
+    ----------
+    key : callable
+        The mapping that returns the value to sort on. Its input will be
+        a tuple that contains a single data point for each source.
+    reverse : boolean value that indicates whether the sort order should
+        be reversed.
+
+    """
+    def __init__(self, key, reverse=False):
+        self.key = key
+        self.reverse = reverse
+
+    def __call__(self, x):
+        values = [self.key(i) for i in zip(*x)]
+        indices = [i for (v, i) in
```

Comment:
Sorry of being picky, but sorting indices is not really Pythonic. In Python copying a reference to an object, i.e. `a=b`, is always a lightweight operation regardless whether the object is an `int` or a huge matrix. Therefore, I would prefer something like this:

```python
def __call__(self, batch):
    with_keys = [(example, self.key(example)) for example in zip(*batch)]
    with_keys.sort(key=lambda _, key : key)
    return [example for example, _ in with_keys]
```

Another thing is precomputing keys. In general it is a nice thing to do, but I quickly checked and seems like `sorted` is smart enough to do it on its own. I could not find it in the docs explicitly, but I would trust that such a function must be optimized. That simplifies our job to the following:

```python
def __call__(self, batch):
    return list(sorted(zip(*batch), key=self.key, reverse=self.reverse))
```

Did I miss anything?